src/operators/TimeEvolutionOperator.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+ * MRCPP, a numerical library based on multiresolution analysis and
+ * the multiwavelet basis which provide low-scaling algorithms as well as
+ * rigorous error control in numerical computations.
+ * Copyright (C) 2021 Stig Rune Jensen, Jonas Juselius, Luca Frediani and contributors.
+ *
+ * This file is part of MRCPP.
+ *
+ * MRCPP is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * MRCPP is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with MRCPP.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * For information on the complete list of contributors to MRCPP, see:
+ * <https://mrcpp.readthedocs.io/>
+ */
+
+#include "TimeEvolutionOperator.h"
+//#include "MRCPP/MWOperators"
+
+
+#include "core/InterpolatingBasis.h"
+#include "core/LegendreBasis.h"
+
+#include "functions/Gaussian.h"
+#include "functions/GaussExp.h"
+
+#include "treebuilders/CrossCorrelationCalculator.h"
+#include "treebuilders/OperatorAdaptor.h"
+#include "treebuilders/SplitAdaptor.h"
+#include "treebuilders/TreeBuilder.h"
+#include "treebuilders/grid.h"
+#include "treebuilders/project.h"
+
+#include "trees/BandWidth.h"
+#include "trees/FunctionTreeVector.h"
+#include "trees/OperatorTree.h"
+
+#include "utils/Printer.h"
+#include "utils/Timer.h"
+#include "utils/math_utils.h"
+
+#include "treebuilders/TimeEvolution_CrossCorrelationCalculator.h"
+
+#include <vector>
+
+
+namespace mrcpp {
+
+
+/** @brief A constructor for TimeEvolutionOperator class.
+ *
+ * @param[in] mra: MRA.
+ * @param[in] prec: precision.
+ * @param[in] time: the time moment (step).
+ * @param[in] finest_scale: the operator constructed down to this scale.
+ * @param[in] imaginary: defines the real (faulse) or imaginary (true) part of the semigroup.
+ * @param[in] max_Jpower: maximum amount of power integrals used.
+ *
+ * @details Constructs either real or imaginary part of the Schrodinger semigroup at a given time moment.
+ * 
+ * 
+ * 
+ */
+template <int D>
+TimeEvolutionOperator<D>::TimeEvolutionOperator
+(const MultiResolutionAnalysis<D> &mra, double prec, double time, int finest_scale, bool imaginary, int max_Jpower)
+    : ConvolutionOperator<D>(mra, mra.getRootScale(), -10)   //One can use ConvolutionOperator instead as well
+{
+    int oldlevel = Printer::setPrintLevel(0);
+    this->setBuildPrec(prec);
+    
+    SchrodingerEvolution_CrossCorrelation cross_correlation(30, mra.getOrder(), mra.getScalingBasis().getScalingType() );
```

Comment:
`crossCorrelation`

Reply:
That's an object and the name is given according to the mrcpp convention.
Check this:
private:
    FunctionData func_data;
    mrcpp::SharedMemory *shared_mem_re;
    mrcpp::SharedMemory *shared_mem_im;
    mrcpp::FunctionTree<3> *re; ///< Real part of function
    mrcpp::FunctionTree<3> *im; ///< Imaginary part of function

    void flushFuncData() {
        this->func_data.real_size = 0;
        this->func_data.imag_size = 0;
        if (this->re != nullptr) {
            this->func_data.real_size = this->re->getNChunksUsed();
            flushMRAData(this->re->getMRA());
        }
        if (this->im != nullptr) {
            this->func_data.imag_size = this->im->getNChunksUsed();
            flushMRAData(this->im->getMRA());
        }
    }

    void flushMRAData(const mrcpp::MultiResolutionAnalysis<3> &mra) {
        const auto &box = mra.getWorldBox();
        this->func_data.type = mra.getScalingBasis().getScalingType();
        this->func_data.order = mra.getOrder();
        this->func_data.depth = mra.getMaxDepth();
        this->func_data.scale = box.getScale();
        this->func_data.boxes[0] = box.size(0);
        this->func_data.boxes[1] = box.size(1);
        this->func_data.boxes[2] = box.size(2);
        this->func_data.corner[0] = box.getCornerIndex().getTranslation(0);
        this->func_data.corner[1] = box.getCornerIndex().getTranslation(1);
        this->func_data.corner[2] = box.getCornerIndex().getTranslation(2);
    }